### PR TITLE
Enum versioning

### PR DESCRIPTION
**Pull request by lpetre**
Adding the ability to change the underlying code for an enum and provide an upgrade path to modify the serialized enum data while inflating.

TODO:
- [x] new type cache data for enums
- [x] upgrade path for enums
- [x] unit tests for new enum serialization
- [x] update readme
- [x] change old enum inflate to behave like we're inflating version zero to get upgrade ability
- [x] flush out getEnumParameterCount for all platforms

### CODE REVIEW

File: serialization/Inflater.hx [hunk 1]
```diff
@@ -493,6 +525,130 @@ class Inflater {
...
+    }
+  }
+
+  function inflateEnumInfo() : InflatedEnum {
+    var info : InflatedEnum = new InflatedEnum();
+    info.name = unserialize();
+    if( stream.readByte() != ":".code ) {
+      throw "Invalid type format (cls)";
+    }
+    info.type = Type.resolveEnum(info.name);
```
Comment (dogles) [inline serialization/Inflater.hx]: This should be a nice performance boost: Type.resolveEnum isn't fast, and we were previously calling it for every enum value.

File: serialization/internal/TypeUtils.hx [hunk 2]
```diff
@@ -212,4 +212,22 @@ class TypeUtils
 
     return null;
   }
+
+  public static function getEnumParameterCount(enm:Enum<Dynamic>, constructor:String) : Int {
+    #if neko
```
Comment (dogles) [inline serialization/internal/TypeUtils.hx]: Still need to fill these in

File: serialization/Deflater.hx [hunk 3]
```diff
@@ -128,7 +147,7 @@ class Deflater {
   **/
   public var useEnumIndex(default, null) : Bool;
 
-  var thash : StringMap<Array<DeflatedClass>>;
+  var thash : StringMap<Array<Dynamic>>;
```
Comment (dogles) [inline serialization/Deflater.hx]: Can we make this an array of &lt;BaseClass&gt;? Array<Dynamic> goes through reflection for field access on some targets. (Not sure why)

File: serialization/internal/TypeUtils.hx (hunk 2 above)
Comment (lpetre) [inline serialization/internal/TypeUtils.hx]: I ended up using an actual enum value instance to get the parameter count. This is using code lifted from the old hxSerialize for enums. I'm concerned that the counts might change depending on optional enum value parameters.

**Pull request merged**